### DISCUSSION TITLE
How to stretch to nearest or interpolate inbetween pixels in a logarithmic audio spectrum converted to a picture in Visual Studio using C# Console

### DISCUSSION THREAD

**Discussion by Winston-17** [SixLabors/ImageSharp#2820] [Q&A]
I'm trying to program a code to convert an audio to an image spectrogram in the scale of logarithmic, but black lines come up. I search for help, I ask 4 different AI chatbots for help, and though it gave me some correct code, it gave wrong or broken code with this problem, Microsoft's Forum told me to go here, and running out of options, I have no choice but to ask for help here. Here is my code;
`using System;
using System.Drawing;
using System.Drawing.Imaging;
using System.Linq;
using NAudio.Wave;
using MathNet.Numerics;
using MathNet.Numerics.IntegralTransforms;
using SixLabors.ImageSharp;
using SixLabors.ImageSharp.PixelFormats;
using SixLabors.ImageSharp.Processing;

class Program
{
    static void Main(string[] args)
    {
        
        string filePath = @"C:\Users\seven\Music\sound.wav";
        int fftLength = 4096; // FFT length

string spectrogramPath = @"C:\Users\seven\Music\spectrogram.bmp"; // Path for the spectrogram image
        string modifiedSpectrogramPath = @"C:\Users\seven\Music\modified_spectrogram.bmp"; // Path for modified spectrogram

using (var reader = new AudioFileReader(filePath))
        {
            var samples = new float[reader.Length / sizeof(float)];
            reader.Read(samples, 0, samples.Length);

var spectrogram = GetLogSpectrogram(samples, reader.WaveFormat.SampleRate, fftLength);

SaveSpectrogramImage(spectrogram, fftLength / 2, spectrogramPath);
        }

ScalePart(@"C:\Users\seven\Music\spectrogram.bmp", modifiedSpectrogramPath);
    }

static double[,] GetLogSpectrogram(float[] samples, int sampleRate, int fftLength)
    {
        int numSpectrogramRows = fftLength / 2;
        int numSpectrogramColumns = (samples.Length / fftLength) * 2;
        double[,] spectrogram = new double[numSpectrogramRows, numSpectrogramColumns];

for (int i = 0; i < numSpectrogramColumns; i++)
        {
            int offset = i * fftLength / 2;
            if (offset + fftLength >= samples.Length)
                break;

var buffer = samples.Skip(offset).Take(fftLength).Select(s => new Complex32(s, 0)).ToArray();
            Fourier.Forward(buffer, FourierOptions.NoScaling);

for (int j = 0; j < numSpectrogramRows; j++)
            {
                spectrogram[j, i] = 20 * Math.Log10(Math.Max(buffer[j].Magnitude, 1e-10));
            }
        }

return spectrogram;
    }

static void SaveSpectrogramImage(double[,] spectrogram, int height, string outputPath)
    {
        int width = spectrogram.GetLength(1);
        using (Bitmap bitmap = new Bitmap(width, height, PixelFormat.Format24bppRgb))
        {
            double max = spectrogram.Cast<double>().Max();
            double min = spectrogram.Cast<double>().Min();
            double gamma = 3;

for (int x = 0; x < width; x++)
            {
                for (int y = 0; y < height; y++)
                {
                    int value = (int)((spectrogram[y, x] - min) / (max - min) * 255);
                    value = (int)(255 * Math.Pow(value / 255.0, gamma)); // Apply gamma correction
                    System.Drawing.Color color = System.Drawing.Color.FromArgb(value, value, value);
                    bitmap.SetPixel(x, height - y - 1, color);
                }
            }

bitmap.Save(outputPath, ImageFormat.Bmp);
        }
    }

static void ScalePart(string inputPath, string outputPath)
    {
        using (var image = SixLabors.ImageSharp.Image.Load<Rgba32>(inputPath))
        {
            int originalHeight = image.Height;

using (var outputImage = new Image<Rgba32>(image.Width, originalHeight))
            {
                for (int x = 0; x < image.Width; x++)
                {
                    for (int y = 0; y < originalHeight; y++)
                    {
                        double logScaleY = Math.Log(y + 1) / Math.Log(originalHeight + 1); // Normalized logarithm
                        int newY = (int)(logScaleY * (originalHeight - 1)); // Transform Y to new position

if (newY < originalHeight)
                        {
                            var color = image[x, originalHeight - y - 1]; // The image is flipped for coordinate compatibility
                            outputImage[x, newY] = color;
                        }
                    }
                }
                   outputImage.Save(outputPath);
            }
        }
    }
}`

![image](https://github.com/user-attachments/assets/9f0e37eb-5df3-474f-b28c-19feb4458b5e)
The black bars at the top I want removed, either by interpolating inbetween lines or stretching the nearest pixel vertically.

AI was giving me such a hard time, I asked Microsoft at after 1:45AM this morning for help after almost a day and a half (give or take the breaks) barely getting it to work at all, even with an alternative, FFMPEG. Now, it has been closer to 2 Days.

This is possible, right? Because even Image Editors can pixelate or interpolate pictures, and warp them in such ways. What is the full code (In this message and what I am trying to achieve)?

If you want to know, I am trying to convert a sound to a logarithmic frequency scale in an image, flip it vertically, redo the frequency, and convert it back to audio, so the pitch is reversed. Many AIs said I could shift the pitch, or resample, or reverse, or flip linearly (Hertz), or use Invert to flip the waveform, but that is not what I want. I basically want to invert the pitch.

These lines might result in additional weird sound, but I want to remove these black stripes and just flip it logarithmically.

Help would be appreciated. If this is the wrong category, please move it to the appropriate category and give me a link to the new path. If someone asked the same thing, please link me there before closing, but I doubt someone asked this question because the AIs programming might not have what I am asking, because nobody is asking.

Thank you for helping.

Quote Mark; _Because reverse is slower than forward; it is speed in the negative._

**Comment by Winston-17** [comment]


**Reply by tiesont** [reply]
If you want support, purchasing a license is a great way to accomplish that task. Otherwise, I think you need to remember that you're posting on the discussion board of an OSS software project that owes you exactly _nothing_, and is unlikely to respond based on the tone of this comment and 

> DO NOT BLOCK/BAN MY QUESTION. THAT WILL MAKE ME CROSS.

**Reply by antonfirsov** [reply]
@Winston-17 also, it would greatly help if you could remove the rant from your post, focus on the the problem description, making it much cleaner and also as minimal as possible. Also, format the code in the markdown correctly and add comments with detailed explanation of the various parts. In it's current form it's very hard to follow, I hardly imagine anyone from the community would want to engage with it in their *free time*.

**Reply by JimBobSquarePants** [reply]
@tiesont Just to note. Purchasing a commercial license does not actually guarantee support. The software licenses is explicitly declared as-is. 

@Winston-17 I very strongly suggest you follow @antonfirsov advice. Your behavior here has been a sever violation of the [code of conduct](https://github.com/SixLabors/ImageSharp?tab=readme-ov-file#code-of-conduct) and is absolutely unacceptable.

**Reply by Winston-17** [reply]
```c#
using System;
using System.Drawing;
using System.Drawing.Imaging;
using System.Linq;
using NAudio.Wave;
using MathNet.Numerics;
using MathNet.Numerics.IntegralTransforms;
using SixLabors.ImageSharp;
using SixLabors.ImageSharp.PixelFormats;
using SixLabors.ImageSharp.Processing;

class Program
{
    static void Main(string[] args)
    {
        
        string filePath = @"sound.wav";
        int fftLength = 4096; // FFT length

        
        string spectrogramPath = @"spectrogram.bmp"; // Path for the spectrogram image
        string modifiedSpectrogramPath = @"modified_spectrogram.bmp"; // Path for modified spectrogram

        using (var reader = new AudioFileReader(filePath))
        {
            var samples = new float[reader.Length / sizeof(float)];
            reader.Read(samples, 0, samples.Length);

            
            var spectrogram = GetLogSpectrogram(samples, reader.WaveFormat.SampleRate, fftLength);

            
            SaveSpectrogramImage(spectrogram, fftLength / 2, spectrogramPath);
        }

        
        ScalePart(@"C:\Users\seven\Music\spectrogram.bmp", modifiedSpectrogramPath);
    }

    static double[,] GetLogSpectrogram(float[] samples, int sampleRate, int fftLength)
    {
        int numSpectrogramRows = fftLength / 2;
        int numSpectrogramColumns = (samples.Length / fftLength) * 2;
        double[,] spectrogram = new double[numSpectrogramRows, numSpectrogramColumns];

        for (int i = 0; i < numSpectrogramColumns; i++)
        {
            int offset = i * fftLength / 2;
            if (offset + fftLength >= samples.Length)
                break;

            var buffer = samples.Skip(offset).Take(fftLength).Select(s => new Complex32(s, 0)).ToArray();
            Fourier.Forward(buffer, FourierOptions.NoScaling);

            for (int j = 0; j < numSpectrogramRows; j++)
            {
                spectrogram[j, i] = 20 * Math.Log10(Math.Max(buffer[j].Magnitude, 1e-10));
            }
        }

        return spectrogram;
    }

    static void SaveSpectrogramImage(double[,] spectrogram, int height, string outputPath)
    {
        int width = spectrogram.GetLength(1);
        using (Bitmap bitmap = new Bitmap(width, height, PixelFormat.Format24bppRgb))
        {
            double max = spectrogram.Cast<double>().Max();
            double min = spectrogram.Cast<double>().Min();
            double gamma = 3;

            for (int x = 0; x < width; x++)
            {
                for (int y = 0; y < height; y++)
                {
                    int value = (int)((spectrogram[y, x] - min) / (max - min) * 255);
                    value = (int)(255 * Math.Pow(value / 255.0, gamma)); // Apply gamma correction
                    System.Drawing.Color color = System.Drawing.Color.FromArgb(value, value, value);
                    bitmap.SetPixel(x, height - y - 1, color);
                }
            }

            bitmap.Save(outputPath, ImageFormat.Bmp);
        }
    }

    static void ScalePart(string inputPath, string outputPath)
    {
        using (var image = SixLabors.ImageSharp.Image.Load<Rgba32>(inputPath))
        {
            int originalHeight = image.Height;

            using (var outputImage = new Image<Rgba32>(image.Width, originalHeight))
            {
                for (int x = 0; x < image.Width; x++)
                {
                    for (int y = 0; y < originalHeight; y++)
                    {
                        double logScaleY = Math.Log(y + 1) / Math.Log(originalHeight + 1);
                        int newY = (int)(logScaleY * (originalHeight - 1));

                        if (newY < originalHeight)
                        {
                            var color = image[x, originalHeight - y - 1];
                            outputImage[x, newY] = color;
                        }
                    }
                }
                   outputImage.Save(outputPath);
            }
        }
    }
}
```

The trouble is that the black lines, like in the picture, I want to remove/fix it.
All I ask is how to stretch pixels when converting EXP/LIN/LOG, and...Seriously? Telling me I violated??

👎👎👎👎👎

This makes me furious. I'm not cussing, I'm trying to be polite, I just want to remove the black lines on the top of the spectrum, and is frustrating asking 4 different AI with broken or incorrect code, asking on Microsoft just to get "This tire does not provide assistance for third-party nuget, you can ask SixLabors for relevant support.", then waiting a week here, and then nothing, having to post only to get the "This Is A Violation" stuff. Do you understand??

What do you want me to do, hold my feelings in until I explode of frustration?? I COULD VERY EASILY SWITCH CODE, LIKE I DID PYTHON. I may never use Python, and I may stop using this.

If you people don't want "DO NOT BLOCK/BAN MY QUESTION. THAT WILL MAKE ME CROSS.", stop banning me for no reason or any of that stuff.

I should be right though, it would be nice to untick "Close as...".

I DID NOT CUSS, I DID NOT WANT TO HARM YOU, BUT ACT LIKE THAT AND I THINK _YOU_ SHOULD BE BANNED. FOR THE SAKES OF IT, FINE! I WILL GIVE YOU ONE, _ONE_ CHANCE, I WILL REMOVE THE DEMAND/RANT/COMPLAINT/WHATEVER, AND I DON'T WANT TO SEE "Account Suspended for trying to politely express frustration", OR "You have been banned for trying to be polite but we call it rude", OR STUFF LIKE THAT.

Not even AI blocks out complaints, but at least it tries to help code.

STOP BEING SO SENSITIVE, YOU COULD LOSE USERS, JUST FOR BANNING AN EXPRESSION, "DO NOT BLOCK/BAN MY QUESTION. THAT WILL MAKE ME CROSS."

**Reply by Winston-17** [reply]
Let me remind you;

It is frustrating asking _4 different AI_ with broken or incorrect code, asking on _Microsoft_ just to get "This tire does not provide assistance for third-party nuget, you can ask _SixLabors_ for relevant support.", then I try, waiting a week here, and then nothing, having to post again only to get the "This Is A Violation" stuff.

It is understand or overstand.

Understand: You realize I am frustrated, trying back-and-forth back-to-back is annoying, and you can help me, or...
Overstand: You ban me for ridiculous reasons, not caring about the users and I stop using this code format.

Please help

**Reply by tiesont** [reply]
You were never "banned", so not sure why you keep bringing that up.

**Lack of "cussing" does not a polite conversation make.** Your tone is quite belligerent, and has been since your original question. I have no idea why. Maybe you were having a bad day, and are having another one today? Whatever it is, you need to dial the aggressiveness _waaaaaaay_ back and try again. 

Or, you know, keep doing what you're doing. Whatever floats your boat, friend.

**Reply by ceberttylertech** [reply]
> you can ask SixLabors for relevant support

Are you paying for commercial support?

**Reply by Walms** [reply]
>YOU COULD LOSE USERS

Sometimes this is a good outcome, I suspect that would be the best possible outcome in this case.

**Reply by Winston-17** [reply]
_"You were never "banned", so not sure why you keep bringing that up."_
Oh? No excuse-- Yes I was. On Python, on DigiBirdsOfficial, On SilverLit, On Reddit, On Yahoo! Answers, And I'm sick of this punishment--what you might call "discipline"-- just because I am frustrated. It is just a feeling, we all have feelings. Sometimes, we can't help ourselves, so please just help me.

First of all, again, let me remind you;

It is frustrating asking 4 different AI with broken or incorrect code, asking on Microsoft just to get "This tire does not provide assistance for third-party nuget, you can ask SixLabors for relevant support.", then I try, waiting a week here, and then nothing, having to post again only to get the "This Is A Violation" stuff.

Second of all, Enough with the frustration, and let's try to focus on the positive aspects now, and get this working; How do I remove the black lines? What do I write?

Third of all, there is a glitch when trying to wrap the entire code. Only some is wrapped, the rest shows as part of the normal text. I don't know why, but I did try to repost it and make sure it was in the quote, and it did the same thing.

Last of all, do you need autism to understand?? Ban after ban after ban, it gets frustrating. Of course I'm going to ask to not be banned, and it is still frustrating when people don't listen, because "I have autism", and now your telling me not to? THERE IS NO KNOWN CURE FOR AUTISM!

This is why I didn't want to have to ask on a forum. Do your job, and please, help me. Enough is enough, so please, don't be negative.

**Reply by Winston-17** [reply]
> Sometimes this is a good outcome, I suspect that would be the best possible outcome in this case.

Like Nintendo, YouTube and Google?

**Reply by JimBobSquarePants** [reply]
@Winston-17 

Take a moment here to relax please because, honestly, I'm a little worried about you. Are you telling me here that you have autism? 

I'd love to reach an amicable solution here but it's difficult with the way the conversation is going. 

Here's the official guide to GitHub Flavored Markdown. Maybe that will help you understanding your formatting errors. 

https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code

**Reply by Winston-17** [reply]
_"Take a moment here to relax please because, honestly, I'm a little worried about you. Are you telling me here that you have autism?"_
Is that all I had to say? Yes I do have autism, so please help me, or I'm the closing question as irrelivent.

_"I'm a little worried about you"_?
Out of all people, 99% say it, and don't seem to care. It is a long story, but if you need to hear it...

My Dad took me to the hospital TWICE when I was almost half the age I am now, my car my friend said he'd let me use as a clubhouse was towed away by others when I was in my late teens early 20s, and the trailor I was going to use as a next-door house wasn't worked and fixed on by anyone they said they would; not by my brother, not by my uncle, not my cousin, not by ANYONE, and wound up becoming completely unusable and had to be taken away. I'm 25, NOT 18, and it seems like it is a complete lie when EVEN MY THERIPISTS _AND_ KID SHOWS said, "You're all grown up when you are 18, and you can do anything you want." Even my Dad bullied me as a kid and teen, even when I try to help him. I would be laughing and having fun, and he'd say "Knock it off" or "Shut up". I try to talk like an adult to people, and he tells me "Don't Bother Them", he would mock me when I was upset, I'd try to run away because of his over-negativity, got taken by the police, the littlest thing angers him, like accidents, and coincedently, he cussed at his video games, and this went on 90% of the time up until he died when it was around 2018, and I still have haunted nightmares of my Dad 6 years later. He is lucky, because if he did something that irritated me one more time, I may have punched him. I've whipped the belt on his urn. So if I'm negative, I am sorry, but I honestly believe my Dad taught me wrong by bulling me, demanding me, and only 2-3 years after he died, things seem to somewhat get better. But I still got sick of it, sick of the memories, and tell my brother's daughters, to try to debunk, "Don't belive it when people say you can do WHATEVER YOU WANT when your 18". And if that doesn't sound jealous enough, even my Mom prohibits me crossing the street, in fear I will get hit WHEN I LEARNED, KNOW, AS A FACT, TO LOOK BOTH WAYS LIKE ANY TODDLER, and our neighbor next door VISITS US and even SHE CROSSES THE STREET! And what is wrong with toilet-humor, if humor is funny? Won't you laugh if I say, "Cars can pass gas, and needs a bathroom, but it's waste is called exhaust"? Not DigiBirdsOffical, but even Little Live Pets made Gotta Go Flamingo! SilverLit threats to ban me, so... If I do get it for Christmas, I'm getting a _used copy_ of Rescue Penguin, because if they want to be like that, I'm not giving them MONEY, nope, not worth it! And I swear, my Dad's spirit is haunting me. Even my Dad's Mom and my Aunt blocked me off their facebook page, probably where the negativity came from. Everyday, I argue with my Mom, and it isn't even her fault. IS THAT REALLY ALL YOU NEEDED TO HEAR?? THAT I HAVE AUTISM?? I hope this clears up any confusion about why I am so angry. I tried to learn from shows, other people, other users, friends; but it is getting increasingly clear that it is a load of mud, and almost nobody is helping me. With the opposite happening now compared to Dad back then, like Dad telling me "You got to stop being yourself", and more recently my sister saying "Just be yourself", it is no wonder she threw the F-bomb at my Dad about a decade and a half ago when she was closer to my age. Similarly, I accidently shut the door on a bird, Dad spanks me. After Dad, another bird escapes, I have to grab it's tail and almost throw him back in his cage, and Mom thanked me for it because it was a pain in the neck to get him in his cage.

To put a long story short, I cannot be normal, mostly or all because of my Dad, and I am more recently seeing some friends, except some people on my Dad's side. There are some friends I might still talk to on Dad's side, but definitely not my Dad's Mom or that Aunt. I won't say who's name for privacy, but you get the point.

Now that I said I have Autism, and expressed my bum off, please, what code do I write to remove/replace the black lines? What do I write?

**Comment by Winston-17** [comment]
![image](https://github.com/user-attachments/assets/82691659-eb1f-4ec8-bd56-308165a31b49)
Here is what I want, all I ask for

**Reply by JimBobSquarePants** [reply]
The problem here is that what you are asking for really isn't part of the ImageSharp domain. We provide limited computer vision functionality in addition to common image processing functionality but mostly we provide the low-level APIs which allow you to implement your own algorithms. 

From your sample code (I reformatted the second sample) It's hard for me to understand whether you have issues simply reading the data from the wave file or whether you need a window function to interpolate between logarithmic values or even if integer casting is causing problems. 

Perhaps the code in this library could give you the relevant pointers? Or even provide you the exact functionality you need?

https://github.com/swharden/Spectrogram
